### PR TITLE
impl(otel): tracing for `rest_internal::HttpPayload`

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11227)

### CODE REVIEW

File: google/cloud/internal/tracing_http_payload.cc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  if (!status) {
+    span_->AddEvent(
+        "Read", {{"read.buffer.size", static_cast<std::int64_t>(buffer.size())},
+                 {"read.error", true}});
+    return internal::EndSpan(*span_, std::move(status));
+  }
+  span_->AddEvent(
+      "Read", {{"read.buffer.size", static_cast<std::int64_t>(buffer.size())},
+               {"read.returned.size", static_cast<std::int64_t>(*status)}});
+  if (*status == 0) span_->End();
```
Comment (dbolduc) [inline google/cloud/internal/tracing_http_payload.cc:45]: I would recommend:

```cc
if (*status == 0) return EndSpan(span_, std::move(status));
```

I'd rather we hit this code path in all of the code, even if we decide that we don't want to add attributes in the success case.

https://github.com/googleapis/google-cloud-cpp/blob/fba937ed589c3a118f712722d93e32c1486c235e/google/cloud/internal/opentelemetry.cc#L52-L57

Aside: You might have read that OpenTelemetry suggests leaving the Span status as `kUnset` in the success case. I have been setting them to `kOk`. I am not worried about removing this line of code. I will just need to update a bunch of tests. https://github.com/googleapis/google-cloud-cpp/blob/fba937ed589c3a118f712722d93e32c1486c235e/google/cloud/internal/opentelemetry.cc#L53

File: google/cloud/internal/tracing_http_payload_test.cc [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+  auto impl = MakeMockHttpPayloadSuccess(MockContents());
+  RestRequest request("https://example.com/ignored");
+  auto request_span = MakeSpanHttp(request, "GET");
+
+  auto span = MakeSpanHttpPayload(*request_span);
+  TracingHttpPayload payload(std::move(impl), span);
+  std::vector<char> buffer(16);
+  auto read = [&] { return payload.Read(absl::Span<char>(buffer)); };
+  for (auto s = read(); s && s.value() != 0; s = read()) continue;
```
Comment (dbolduc) [inline google/cloud/internal/tracing_http_payload_test.cc:64]: Lol, 😉

File: google/cloud/internal/tracing_http_payload_test.cc [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  EXPECT_CALL(*impl, Read)
+      .WillOnce(Return(16))
+      .WillOnce(Return(Status(StatusCode::kUnavailable, "try-again")));
+
+  auto span = MakeSpanHttpPayload(*request_span);
+  TracingHttpPayload payload(std::move(impl), span);
+  std::vector<char> buffer(16);
+  auto status = payload.Read(absl::Span<char>(buffer.data(), buffer.size()));
+  ASSERT_STATUS_OK(status);
+  EXPECT_THAT(*status, Eq(16));
```
Comment (dbolduc) [inline google/cloud/internal/tracing_http_payload_test.cc:102]: nit: `EXPECT_EQ` ?

File: google/cloud/internal/rest_opentelemetry_test.cc [hunk 4]
```diff
@@ -66,6 +67,26 @@ TEST(RestOpentelemetry, MakeSpanHttp) {
...
 }
 
+TEST(RestOpentelemetry, MakeSpanHttpPayload) {
+  namespace sc = ::opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+
+  RestRequest request("https://example.com/ignored");
+  auto request_span = MakeSpanHttp(request, "GET");
+
+  auto span = MakeSpanHttpPayload(*request_span);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/rest_opentelemetry_test.cc:77]: Created #11236, I am rushing to get tracing working for a customer case.

File: google/cloud/internal/tracing_http_payload_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/tracing_http_payload_test.cc]: Done.

File: google/cloud/internal/tracing_http_payload.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/tracing_http_payload.cc]: Done

**Pull request merged**